File: ios/LookSeeProto/UI Files/Signup.swift
//
//  Signup.swift
//  LookSeeProto
//
//  Created by Sheenan Ahsan on 2/27/26.
//


import SwiftUI
import Amplify

struct Signup: View {
    @State private var username = ""
    @State private var email = ""
    @State private var password = ""
    @State private var confirmPassword = ""
    @State private var verificationCode = ""
    @State private var isPasswordVisible = false
    @State private var isConfirmPasswordVisible = false
    
    @State private var message = ""
    @State private var isLoading = false
    @State private var showVerification = false
    @State private var isBusinessAccount = false
    
    @EnvironmentObject var vm: AuthViewModel

    var onSignupSuccess: (String) -> Void
    var onGoToLogin: () -> Void
    
    private func isValidPassword(_ pass: String) -> Bool {
        let passwordRegex = "^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])(?=.*[^A-Za-z0-9]).{8,}$"
        return NSPredicate(format: "SELF MATCHES %@", passwordRegex).evaluate(with: pass)
    }
    
    private func hideKeyboard() {
        UIApplication.shared.sendAction(#selector(UIResponder.resignFirstResponder), to: nil, from: nil, for: nil)
    }

    var body: some View {
        ZStack {
            Color(red: 0.06, green: 0.06, blue: 0.10).ignoresSafeArea()
                .onTapGesture { hideKeyboard() }
            
            Circle().fill(Color(red: 0.22, green: 0.49, blue: 1.00).opacity(0.12)).frame(width: 300, height: 300).blur(radius: 60).offset(y: -100)

            ScrollView {
                VStack(spacing: 0) {
                    VStack(spacing: 2) {
                        Image("LookSee_Logo").resizable().scaledToFit().frame(width: 350, height: 300)
                        Text(showVerification ? "Check your email" : "Create your account").font(.subheadline).foregroundStyle(Color.white.opacity(0.4))
                    }
                    .padding(.top, 20)

                    VStack(spacing: 14) {
                        if !showVerification {
                            // MARK: - SIGNUP FORM
                            VStack(alignment: .leading, spacing: 6) {
                                Text("Unique Username").font(.caption).foregroundStyle(Color.white.opacity(0.9))
                                TextField(
                                    "username",
                                    text: $username,
                                    prompt: Text("username").foregroundStyle(Color.white.opacity(0.6))
                                )
                                    .textInputAutocapitalization(.never)
                                    .autocorrectionDisabled(true)
                                    .padding(14).background(Color(red: 0.18, green: 0.18, blue: 0.24)).foregroundStyle(.white).cornerRadius(12)
                                    .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(red: 0.22, green: 0.49, blue: 1.00).opacity(0.3), lineWidth: 0.5)).colorScheme(.dark)
                                    .onChange(of: username) { _, newValue in
                                        username = newValue.lowercased().filter { "abcdefghijklmnopqrstuvwxyz0123456789_".contains($0) }
                                    }
                            }
                            
                            VStack(alignment: .leading, spacing: 6) {
                                Text("Email").font(.caption).foregroundStyle(Color.white.opacity(0.9))
                                TextField(
                                    "Email",
                                    text: $email,
                                    prompt: Text("you@example.com").foregroundStyle(Color.white.opacity(0.6))
                                )
                                    .keyboardType(.emailAddress).autocorrectionDisabled(true).textInputAutocapitalization(.never)
                                    .padding(14).background(Color(red: 0.18, green: 0.18, blue: 0.24)).foregroundStyle(.white).cornerRadius(12)
                                    .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(red: 0.22, green: 0.49, blue: 1.00).opacity(0.3), lineWidth: 0.5)).colorScheme(.dark)
                            }

                            VStack(alignment: .leading, spacing: 6) {
                                Text("Password").font(.caption).foregroundStyle(Color.white.opacity(0.9))
                                HStack(spacing: 10) {
                                    Group {
                                        if isPasswordVisible {
                                            TextField(
                                                "Password",
                                                text: $password,
                                                prompt: Text("Enter your password").foregroundStyle(Color.white.opacity(0.6))
                                            )
                                        } else {
                                            SecureField(
                                                "Password",
                                                text: $password,
                                                prompt: Text("••••••••").foregroundStyle(Color.white.opacity(0.6))
                                            )
                                        }
                                    }
                                    .textContentType(.newPassword)
                                    .textInputAutocapitalization(.never)
                                    .autocorrectionDisabled(true)

                                    Button {
                                        isPasswordVisible.toggle()
                                    } label: {
                                        Image(systemName: isPasswordVisible ? "eye.slash.fill" : "eye.fill")
                                            .foregroundStyle(Color.white.opacity(0.75))
                                            .frame(width: 24, height: 24)
                                    }
                                    .buttonStyle(.plain)
                                    .accessibilityLabel(isPasswordVisible ? "Hide password" : "Show password")
                                }
                                    .padding(14).background(Color(red: 0.18, green: 0.18, blue: 0.24)).foregroundStyle(.white).cornerRadius(12)
                                    .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(red: 0.22, green: 0.49, blue: 1.00).opacity(0.3), lineWidth: 0.5)).colorScheme(.dark)
                                
                                if !password.isEmpty && !isValidPassword(password) {
                                    Text("Requires 8+ chars, 1 uppercase, 1 lowercase, 1 number, and 1 special char.")
                                        .font(.system(size: 11)).foregroundStyle(.red).padding(.leading, 4)
                                }
                            }

                            VStack(alignment: .leading, spacing: 6) {
                                Text("Confirm Password").font(.caption).foregroundStyle(Color.white.opacity(0.9))
                                HStack(spacing: 10) {
                                    Group {
                                        if isConfirmPasswordVisible {
                                            TextField(
                                                "Confirm Password",
                                                text: $confirmPassword,
                                                prompt: Text("Retype your password").foregroundStyle(Color.white.opacity(0.6))
                                            )
                                        } else {
                                            SecureField(
                                                "Confirm Password",
                                                text: $confirmPassword,
                                                prompt: Text("Retype your password").foregroundStyle(Color.white.opacity(0.6))
                                            )
                                        }
                                    }
                                    .textContentType(.newPassword)
                                    .textInputAutocapitalization(.never)
                                    .autocorrectionDisabled(true)

                                    Button {
                                        isConfirmPasswordVisible.toggle()
                                    } label: {
                                        Image(systemName: isConfirmPasswordVisible ? "eye.slash.fill" : "eye.fill")
                                            .foregroundStyle(Color.white.opacity(0.75))
                                            .frame(width: 24, height: 24)
                                    }
                                    .buttonStyle(.plain)
                                    .accessibilityLabel(isConfirmPasswordVisible ? "Hide confirmation password" : "Show confirmation password")
                                }
                                    .padding(14).background(Color(red: 0.18, green: 0.18, blue: 0.24)).foregroundStyle(.white).cornerRadius(12)
                                    .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(red: 0.22, green: 0.49, blue: 1.00).opacity(0.3), lineWidth: 0.5)).colorScheme(.dark)

                                if !confirmPassword.isEmpty && password != confirmPassword {
                                    Text("Passwords do not match.")
                                        .font(.system(size: 11)).foregroundStyle(.red).padding(.leading, 4)
                                }
                            }

                            HStack {
                                VStack(alignment: .leading, spacing: 2) {
                                    Text("Business Account").font(.subheadline).foregroundStyle(.white)
                                    Text("Enables promotion management and video uploads").font(.caption).foregroundStyle(Color.white.opacity(0.4))
                                }
                                Spacer()
                                Toggle("", isOn: $isBusinessAccount).tint(Color(red: 0.22, green: 0.49, blue: 1.00))
                            }
                            .padding(14).background(Color(red: 0.18, green: 0.18, blue: 0.24)).cornerRadius(12).overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(red: 0.22, green: 0.49, blue: 1.00).opacity(0.3), lineWidth: 0.5))
                            
                        } else {
                            // MARK: - VERIFICATION FORM
                            VStack(alignment: .leading, spacing: 6) {
                                Text("Enter 6-Digit Code").font(.caption).foregroundStyle(Color.white.opacity(0.9))
                                TextField(
                                    "Verification Code",
                                    text: $verificationCode,
                                    prompt: Text("123456").foregroundStyle(Color.white.opacity(0.6))
                                )
                                    .keyboardType(.numberPad)
                                    .font(.system(size: 24, weight: .bold, design: .monospaced))
                                    .multilineTextAlignment(.center)
                                    .padding(14).background(Color(red: 0.18, green: 0.18, blue: 0.24)).foregroundStyle(.white).cornerRadius(12)
                                    .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(red: 0.22, green: 0.49, blue: 1.00).opacity(0.3), lineWidth: 0.5)).colorScheme(.dark)
                            }
                        }

                        if !message.isEmpty {
                            Text(message).font(.caption).foregroundStyle(message.contains("successful") || message.contains("verified") ? .green : .red).frame(maxWidth: .infinity, alignment: .leading)
                        }

                        Button {
                            if showVerification { verifyCode() } else { signUp() }
                        } label: {
                            HStack(spacing: 8) {
                                if isLoading {
                                    ProgressView().tint(.white)
                                } else {
                                    Text(showVerification ? "Verify Account" : "Create Account").font(.system(size: 16, weight: .semibold))
                                    Image(systemName: "arrow.right").font(.system(size: 14, weight: .semibold))
                                }
                            }
                            .foregroundStyle(.white).frame(maxWidth: .infinity).padding(.vertical, 16).background(Color(red: 0.22, green: 0.49, blue: 1.00)).cornerRadius(14)
                            .opacity(isLoading || (showVerification ? verificationCode.count < 6 : (email.isEmpty || username.isEmpty || !isValidPassword(password) || confirmPassword.isEmpty || password != confirmPassword)) ? 0.5 : 1)
                        }
                        .disabled(isLoading || (showVerification ? verificationCode.count < 6 : (email.isEmpty || username.isEmpty || !isValidPassword(password) || confirmPassword.isEmpty || password != confirmPassword)))
                        .padding(.top, 16)

                        if !showVerification {
                            Button { onGoToLogin() } label: {
                                HStack(spacing: 0) {
                                    Text("Already have an account? ").foregroundStyle(Color.white.opacity(0.4))
                                    Text("Sign in").foregroundStyle(Color(red: 0.22, green: 0.49, blue: 1.00))
                                }.font(.footnote)
                            }
                            .padding(.top, 8)
                        }
                    }
                    .padding(.horizontal, 28).padding(.bottom, 52)
                }
            }
            .scrollDismissesKeyboard(.interactively)
        }
    }

    // MARK: - AWS Logic
    func signUp() {
        guard password == confirmPassword else {
            message = "Passwords do not match."
            return
        }

        isLoading = true
        message = ""
        let group = isBusinessAccount ? "business-users" : "authenticated-users"
        Task {
            do {
                let result = try await AuthService.shared.signUp(username: email, password: password, email: email, group: group)
                
                if result.isSignUpComplete {
                    // 🚀 FIXED: Memorize username to be processed AFTER login!
                    await MainActor.run { vm.pendingUsernameToSave = username }
                    message = "Account created and verified! Routing to login..."
                    DispatchQueue.main.asyncAfter(deadline: .now() + 1.2) {
                        onSignupSuccess(email)
                    }
                } else {
                    withAnimation { showVerification = true }
                    message = "Code sent! Please check your email."
                }
            } catch {
                message = error.localizedDescription
            }
            isLoading = false
        }
    }
    
    func verifyCode() {
        isLoading = true
        message = ""
        Task {
            do {
                let result = try await Amplify.Auth.confirmSignUp(for: email, confirmationCode: verificationCode)
                if result.isSignUpComplete {
                    // 🚀 FIXED: Memorize username to be processed AFTER login!
                    await MainActor.run { vm.pendingUsernameToSave = username }
                    message = "Verification successful! Routing to login..."
                    DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
                        onSignupSuccess(email)
                    }
                } else {
                    message = "Verification incomplete. Please check the code."
                }
            } catch {
                message = error.localizedDescription
            }
            isLoading = false
        }
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/uifiles/Signup.kt
package looksee.angelll.com.uifiles

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.blur
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.amplifyframework.kotlin.core.Amplify
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import looksee.angelll.com.services.AuthService
import looksee.angelll.com.ui.theme.LookSeeCard
import looksee.angelll.com.viewmodels.AuthViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun Signup(
    vm: AuthViewModel,
    onSignupSuccess: () -> Unit,
    onGoToLogin: () -> Unit
) {
    var email by remember { mutableStateOf("") }
    var username by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var verificationCode by remember { mutableStateOf("") }
    var isBusinessAccount by remember { mutableStateOf(false) }

    var showVerification by remember { mutableStateOf(false) }
    var isLoading by remember { mutableStateOf(false) }
    var message by remember { mutableStateOf("") }

    val coroutineScope = rememberCoroutineScope()

    val isValidPassword = { pass: String ->
        val passwordRegex = "^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])(?=.*[^A-Za-z0-9]).{8,}$".toRegex()
        passwordRegex.matches(pass)
    }

    val signUp: () -> Unit = {
        isLoading = true
        message = ""
        val group = if (isBusinessAccount) "business-users" else "authenticated-users"
        coroutineScope.launch {
            try {
                val result = AuthService.signUp(email, password, email, group)
                if (result.isSignUpComplete) {
                    vm.pendingUsernameToSave = username
                    message = "Account created and verified! Routing to login..."
                    delay(1200)
                    onSignupSuccess()
                } else {
                    showVerification = true
                    message = "Code sent! Please check your email."
                }
            } catch (e: Exception) {
                message = e.localizedMessage ?: "Signup failed."
            } finally {
                isLoading = false
            }
        }
    }

    val verifyCode: () -> Unit = {
        isLoading = true
        message = ""
        coroutineScope.launch {
            try {
                val result = Amplify.Auth.confirmSignUp(email, verificationCode)
                if (result.isSignUpComplete) {
                    vm.pendingUsernameToSave = username
                    message = "Verification successful! Routing to login..."
                    delay(1500)
                    onSignupSuccess()
                } else {
                    message = "Verification incomplete. Please check the code."
                }
            } catch (e: Exception) {
                message = e.localizedMessage ?: "Verification failed."
            } finally {
                isLoading = false
            }
        }
    }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xFF0F0F1A)),
        contentAlignment = Alignment.TopCenter
    ) {
        // Glow effect
        Box(
            modifier = Modifier
                .size(300.dp)
                .offset(y = (-100).dp)
                .background(Brush.radialGradient(listOf(Color(0xFF387DFF).copy(alpha = 0.12f), Color.Transparent)))
                .blur(60.dp)
        )

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 28.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Image(
                    painter = painterResource(id = looksee.angelll.com.R.drawable.looksee_logo),
                    contentDescription = "LookSee Logo",
                    modifier = Modifier.size(350.dp)
                )
                Text(
                    text = if (showVerification) "Check your email" else "Create your account",
                    color = Color.White.copy(alpha = 0.4f),
                    fontSize = 14.sp
                )
            }

            Spacer(modifier = Modifier.height(14.dp))

            if (!showVerification) {
                // SIGNUP FORM
                Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
                    VStackLabel(label = "Unique Username") {
                        OutlinedTextField(
                            value = username,
                            onValueChange = { newValue ->
                                username = newValue.lowercase().filter { it.isLetterOrDigit() || it == '_' }
                            },
                            placeholder = { Text("username", color = Color.Gray) },
                            modifier = Modifier.fillMaxWidth(),
                            singleLine = true,
                            shape = RoundedCornerShape(12.dp),
                            colors = OutlinedTextFieldDefaults.colors(
                                focusedTextColor = Color.White,
                                unfocusedTextColor = Color.White,
                                focusedContainerColor = Color(0xFF2D2D3D),
                                unfocusedContainerColor = Color(0xFF2D2D3D),
                                focusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f),
                                unfocusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f)
                            )
                        )
                    }

                    VStackLabel(label = "Email") {
                        OutlinedTextField(
                            value = email,
                            onValueChange = { email = it },
                            placeholder = { Text("you@example.com", color = Color.Gray) },
                            modifier = Modifier.fillMaxWidth(),
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
                            singleLine = true,
                            shape = RoundedCornerShape(12.dp),
                            colors = OutlinedTextFieldDefaults.colors(
                                focusedTextColor = Color.White,
                                unfocusedTextColor = Color.White,
                                focusedContainerColor = Color(0xFF2D2D3D),
                                unfocusedContainerColor = Color(0xFF2D2D3D),
                                focusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f),
                                unfocusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f)
                            )
                        )
                    }

                    VStackLabel(label = "Password") {
                        OutlinedTextField(
                            value = password,
                            onValueChange = { password = it },
                            placeholder = { Text("••••••••", color = Color.Gray) },
                            modifier = Modifier.fillMaxWidth(),
                            visualTransformation = PasswordVisualTransformation(),
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
                            singleLine = true,
                            shape = RoundedCornerShape(12.dp),
                            colors = OutlinedTextFieldDefaults.colors(
                                focusedTextColor = Color.White,
                                unfocusedTextColor = Color.White,
                                focusedContainerColor = Color(0xFF2D2D3D),
                                unfocusedContainerColor = Color(0xFF2D2D3D),
                                focusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f),
                                unfocusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f)
                            )
                        )
                        if (password.isNotEmpty() && !isValidPassword(password)) {
                            Text(
                                "Requires 8+ chars, 1 uppercase, 1 lowercase, 1 number, and 1 special char.",
                                color = Color.Red,
                                fontSize = 11.sp,
                                modifier = Modifier.padding(start = 4.dp, top = 4.dp)
                            )
                        }
                    }

                    // Business Account Toggle
                    LookSeeCard {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                                Text("Business Account", color = Color.White, fontSize = 15.sp)
                                Text("Enables promotion management and video uploads", color = Color.White.copy(alpha = 0.4f), fontSize = 12.sp)
                            }
                            Switch(
                                checked = isBusinessAccount,
                                onCheckedChange = { isBusinessAccount = it },
                                colors = SwitchDefaults.colors(checkedThumbColor = Color.White, checkedTrackColor = Color(0xFF007AFF))
                            )
                        }
                    }
                }
            } else {
                // VERIFICATION FORM
                VStackLabel(label = "Enter 6-Digit Code") {
                    OutlinedTextField(
                        value = verificationCode,
                        onValueChange = { if (it.length <= 6) verificationCode = it },
                        placeholder = { Text("123456", color = Color.Gray, modifier = Modifier.fillMaxWidth(), textAlign = TextAlign.Center) },
                        modifier = Modifier.fillMaxWidth(),
                        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                        singleLine = true,
                        textStyle = LocalTextStyle.current.copy(textAlign = TextAlign.Center, fontSize = 24.sp, fontWeight = FontWeight.Bold),
                        shape = RoundedCornerShape(12.dp),
                        colors = OutlinedTextFieldDefaults.colors(
                            focusedTextColor = Color.White,
                            unfocusedTextColor = Color.White,
                            focusedContainerColor = Color(0xFF2D2D3D),
                            unfocusedContainerColor = Color(0xFF2D2D3D),
                            focusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f),
                            unfocusedBorderColor = Color(0xFF387DFF).copy(alpha = 0.3f)
                        )
                    )
                }
            }

            if (message.isNotEmpty()) {
                Text(
                    message,
                    color = if (message.contains("successful") || message.contains("verified")) Color.Green else Color.Red,
                    fontSize = 12.sp,
                    modifier = Modifier.fillMaxWidth().padding(top = 8.dp)
                )
            }

            Spacer(modifier = Modifier.height(24.dp))

            val canSubmit = if (showVerification) verificationCode.length >= 6 else (email.isNotEmpty() && username.isNotEmpty() && isValidPassword(password))

            Button(
                onClick = { if (showVerification) verifyCode() else signUp() },
                modifier = Modifier.fillMaxWidth().height(56.dp),
                shape = RoundedCornerShape(16.dp),
                colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF007AFF), disabledContainerColor = Color(0xFF007AFF).copy(alpha = 0.5f)),
                enabled = canSubmit && !isLoading
            ) {
                if (isLoading) {
                    CircularProgressIndicator(color = Color.White, modifier = Modifier.size(24.dp))
                } else {
                    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        Text(if (showVerification) "Verify Account" else "Create Account", fontSize = 16.sp, fontWeight = FontWeight.SemiBold)
                        Icon(Icons.AutoMirrored.Filled.ArrowForward, contentDescription = null, modifier = Modifier.size(16.dp))
                    }
                }
            }

            if (!showVerification) {
                TextButton(onClick = onGoToLogin, modifier = Modifier.padding(top = 8.dp)) {
                    Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                        Text("Already have an account?", color = Color.White.copy(alpha = 0.4f), fontSize = 12.sp)
                        Text("Sign in", color = Color(0xFF007AFF), fontSize = 12.sp)
                    }
                }
            }

            Spacer(modifier = Modifier.height(52.dp))
        }
    }
}

@Composable
fun VStackLabel(label: String, content: @Composable () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Text(label, color = Color.White.copy(alpha = 0.5f), fontSize = 12.sp)
        content()
    }
}
